plugins/amdgpu/amdgpu_plugin.c

```diff
@@ -621,6 +626,83 @@ void *dump_bo_contents(void *_thread_data)
         return NULL;
 };
 
+int check_hsakmt_shared_mem(uint64_t *shared_mem_size, uint32_t *shared_mem_magic)
+{
+	int ret, fd;
+	struct stat st;
+	uint32_t *shared_mem_ptr;
+
+	ret = stat(HSAKMT_SHM_PATH, &st);
+	if (ret) {
+		*shared_mem_size = 0;
+		return 0;
+	}
+
+	fd = shm_open(HSAKMT_SHM, O_RDONLY, 0);
+	if (fd < 0) {
+		*shared_mem_size = 0;
+		return 0;
+	}
+
+	shared_mem_ptr = mmap(0, st.st_size, PROT_READ, MAP_SHARED, fd, 0);
+	if (shared_mem_ptr == MAP_FAILED) {
+		pr_perror("amdgpu_plugin: Failed to mmap shared mem %s\n", HSAKMT_SHM);
+		return -EACCES;
+	}
+
+	*shared_mem_size = st.st_size;
+
+	/* First 4 bytes of shared file is the magic */
+	memcpy(shared_mem_magic, shared_mem_ptr, sizeof(*shared_mem_magic));
+	plugin_log_msg("amdgpu_plugin: Shared mem magic:0x%x\n", *shared_mem_magic);
+
+	close(fd);
+	return 0;
+}
+
+int restore_hsakmt_shared_mem(const uint64_t shared_mem_size, const uint32_t shared_mem_magic)
+{
+	int ret, fd;
+	struct stat st;
+	uint32_t *shared_mem_ptr;
+	sem_t *sem = SEM_FAILED;
+
+	if (!shared_mem_size)
+		return 0;
+
+	if (!stat(HSAKMT_SHM_PATH, &st)) {
+		pr_debug("amdgpu_plugin: %s already exists\n", HSAKMT_SHM_PATH);
+	} else {
+		pr_info("Warning:%s was missing. Re-creating new file but we may lose perf counters\n",
+										HSAKMT_SHM_PATH);
+		fd = shm_open(HSAKMT_SHM, O_CREAT | O_RDWR, 0666);
+
+		ret = ftruncate(fd, shared_mem_size);
+		if (ret < 0) {
+			pr_err("amdgpu_plugin: Failed to truncate shared mem %s\n", HSAKMT_SHM);
+			close(fd);
+			return -EINVAL;
+		}
+
+		shared_mem_ptr = mmap(0, shared_mem_size, PROT_WRITE, MAP_SHARED, fd, 0);
+		if (shared_mem_ptr == MAP_FAILED) {
+			pr_perror("amdgpu_plugin: Failed to mmap shared mem %s\n", HSAKMT_SHM);
+			close(fd);
+			return -EACCES;
+		}
+		memcpy(shared_mem_ptr, &shared_mem_magic, sizeof(shared_mem_magic));
+		close(fd);
+	}
+
+	sem = sem_open(HSAKMT_SEM, O_CREAT, 0666, 0);
+	if (sem == SEM_FAILED) {
+		pr_perror("Failed to create %s\n", HSAKMT_SEM);
+		return -EACCES;
```

Comment:
It returns SEM_FAILED with set error so just return -1 as pr_perror takes care of showing actual error

Reply:
This will not affect behavior. Caller just checks for non-zero.

Reply:
This is just a nit and I can take care of this while applying this PR.  @fxkamd if you don't have any objections to this change, I think its good to get included in our staging.

Reply:
We're making up error codes in a number of places. If we wanted to return the real error code, we'd return -errno. But the caller ignores this anyway. So it really doesn't matter.